File: app/src/main/java/com/example/boulderjournal/RouteAdapter.java
package com.example.boulderjournal;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.boulderjournal.Utils.Utilities;
import com.example.boulderjournal.data.RouteEntry;

import java.util.List;


public class RouteAdapter extends RecyclerView.Adapter<RouteAdapter.RouteViewHolder> {

    private Context mContext;
    private List<RouteEntry> mRouteEntries;
    final private ItemClickListener mItemClickListener;

    public RouteAdapter(Context mContext, ItemClickListener listener) {
        this.mContext = mContext;
        mItemClickListener = listener;
    }


    @NonNull
    @Override
    public RouteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(mContext)
                .inflate(R.layout.route_input_layout, parent, false);

        return new RouteViewHolder(view);
    }

    @Override
    public void onBindViewHolder( RouteViewHolder holder, int position) {

        RouteEntry routeEntry = mRouteEntries.get(position);
        String routeName = routeEntry.getRouteName();
        holder.routeTitleMain.setText(routeName);
        String routeColor = routeEntry.getRouteColour();
        int routeColourInt = Utilities.getColor(routeColor, mContext);
        holder.routeColourSwatch.setBackgroundColor(routeColourInt);

    }




    @Override
    public int getItemCount() {
        if (mRouteEntries == null) {
            return 0;
        }
        int count = mRouteEntries.size();
        return count;
    }



    public void setRoutes(List<RouteEntry> routeEntries){
        mRouteEntries = routeEntries;
        notifyDataSetChanged();
    }

    public List<RouteEntry> getRoutes(){
        return mRouteEntries;

    }

    public interface ItemClickListener {
        void onItemClickListener(int itemId);
    }



    class RouteViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        TextView routeTitleMain;
        View routeColourSwatch;

        public RouteViewHolder( View itemView) {
            super(itemView);
            routeTitleMain = (TextView)itemView.findViewById(R.id.route_name_main);
            routeColourSwatch = (View)itemView.findViewById(R.id.color_swatch_main);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int elementId = mRouteEntries.get(getAdapterPosition()).getId();
            mItemClickListener.onItemClickListener(elementId);
        }
    }


}

File: app/src/main/java/com/example/boulderjournal/RouteAdapter.kt
package com.example.boulderjournal

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.boulderjournal.Utils.Utilities
import com.example.boulderjournal.data.RouteEntry
import com.example.boulderjournal.databinding.RouteInputLayoutBinding


class RouteAdapter(private val mItemClickListener: ItemClickListener) :  ListAdapter<RouteEntry, RouteAdapter.RouteViewHolder>(RouteEntryDiffCallBack()){

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RouteViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val binding = RouteInputLayoutBinding.inflate(layoutInflater,parent, false)
        return RouteViewHolder(binding, parent.context)
    }

    override fun onBindViewHolder(holder: RouteViewHolder, position: Int) {
        val routeEntry = getItem(position)
        holder.bindRoute(routeEntry)
    }

    interface ItemClickListener {
        fun onItemClickListener(itemId: Int)
    }

    fun getRouteByPosition(position: Int) : RouteEntry{
        return getItem(position)
    }

    inner class RouteViewHolder(val binding: RouteInputLayoutBinding, val context: Context) : RecyclerView.ViewHolder(binding.root){


        fun bindRoute(routeEntry: RouteEntry) {
            val routeName = routeEntry.routeName
            binding.routeNameMain.text = routeName
            val routeColor = routeEntry.routeColour
            val routeColourInt = Utilities.getColor(routeColor, context)
            binding.colorSwatchMain.setBackgroundColor(routeColourInt)
        }
    }

    class RouteEntryDiffCallBack() :
            DiffUtil.ItemCallback<RouteEntry>(){
        override fun areItemsTheSame(oldItem: RouteEntry, newItem: RouteEntry): Boolean {
            return  oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: RouteEntry, newItem: RouteEntry): Boolean {
            return oldItem.equals(newItem)
        }


    }

}
